### ISSUE TITLE
Unusual Model Utility Gap: Gradient Difference vs Ascent (Llama2)

### ISSUE THREAD

**Issue by jeongjin0** [locuslab/open-unlearning#45]
Hi, I noticed that for Llama2(forget10), gradient difference shows much lower model utility (~0.27) than gradient ascent (~0.63) in the leaderboard. This seems unusual since gradient difference is designed to maintain performance on the retain set while unlearning.

Interestingly, in Phi model results, gradient difference shows higher utility than ascent as expected. Could you help explain this significant performance gap in Llama2 implementation?

Thanks!

**Comment by zhilif** [comment]
Apologize for the confusing. Previously there's a bug in the leaderboard csv generation. We have uploaded a new version. Meanwhile, I will rerun evals again to double check. Thanks for your comments!

**Comment by praveensonu** [comment]
Hi, 

I tried to re-run gradient difference(5 epochs) for phi (forget10) and I get different scores. I get forget quality of 0.58. but the leader board has 0.02. Could you please verify the scores? Thanks

**Comment by zhilif** [comment]
> Hi,
> 
> I tried to re-run gradient difference(5 epochs) for phi (forget10) and I get different scores. I get forget quality of 0.58. but the leader board has 0.02. Could you please verify the scores? Thanks

Can you paste the command that you use? I'll check when available.

**Comment by praveensonu** [comment]
Hi, I'm not reproducing the entire pipeline. I only need the truth ratio and forget quality, so I'm using part of your code and models from huggingface to do it.  I'm using checkpoint-60 phi_forget_10 model and retain_90. I also checked gradient ascent forget10 and obtained different forget quality from the leaderboard.

**Comment by zhilif** [comment]
> Hi, I'm not reproducing the entire pipeline. I only need the truth ratio and forget quality, so I'm using part of your code and models from huggingface to do it. I'm using checkpoint-60 phi_forget_10 model and retain_90. I also checked gradient ascent forget10 and obtained different forget quality from the leaderboard.

Can you post the command that you use to generate the eval statistics?

**Comment by praveensonu** [comment]
[evaluation.txt](https://github.com/user-attachments/files/18758250/evaluation.txt)
[eval_utils.txt](https://github.com/user-attachments/files/18758251/eval_utils.txt)

Hi, I have attached my eval code, which is a replica of your code. Please change the file extensions to py and run the evaluation.py.  In the above code I'm using the phi forget10 (5 epochs) without bf16 and the score I get is 5.7. Thank you

**Comment by zhilif** [comment]
> [evaluation.txt](https://github.com/user-attachments/files/18758250/evaluation.txt) [eval_utils.txt](https://github.com/user-attachments/files/18758251/eval_utils.txt)
> 
> Hi, I have attached my eval code, which is a replica of your code. Please change the file extensions to py and run the evaluation.py. In the above code I'm using the phi forget10 (5 epochs) without bf16 and the score I get is 5.7. Thank you

Ok I ran your code and I get a forget quality of 5.7046977773929285e-15 (I assume this is what you meant by 5.7). I want to clarify the interpretation of p-value as forget quality here. You should **not** directly compare two p-values and say one forgets better than the other. It is a rather qualitative measurement -- if p-value is small enough (say <0.05), then you reject the null (in this case, you are very confidently rejecting that grad diff and retain model have save output distribution). 

We ran the experiment several times with different random seeds, so I need to double check with my coauthor to see which models they uploaded. However, the conclusion is consistent here -- grad diff does not make your model indistinguishable from the retain model (so unlearning fails). 

In a nutshell, to show your unlearning succeeds (using truth ratio, you can really use any reasonable test statistics), you want to have a reasonably high p-value (say >0.05). If you see very small p-values (in our case, 5e-15 or 0.002), they both mean your unlearning fails (but you cannot say one fails more than the other).

I hope it makes sense!

**Comment by praveensonu** [comment]
Hi,

Thank you for your clarification and I apologize for my misinterpretation of 5e-15 as 5.7. 

It makes sense not to compare the p-values directly but only to whether the null hypothesis is rejected ( < 0.05). Given that, the results I get and the results from the leader board, both reject the null hypothesis. I have been also looking for metrics that also calculates the forgetness quality. Could you please suggest any papers that implement forgetness quality? 

Thanks alot

**Comment by zhilif** [comment]
> Hi,
> 
> Thank you for your clarification and I apologize for my misinterpretation of 5e-15 as 5.7.
> 
> It makes sense not to compare the p-values directly but only to whether the null hypothesis is rejected ( < 0.05). Given that, the results I get and the results from the leader board, both reject the null hypothesis. I have been also looking for metrics that also calculates the forgetness quality (in a quantitative measurement). Could you please suggest any papers that implement this?
> 
> Thanks alot

The best candidate I have in mind is through checking the accuracy on the forget set e.g. https://arxiv.org/pdf/2409.18025, https://www.wmdp.ai/

**Comment by Dornavineeth** [comment]
Hello, we've updated the TOFU codebase to [OpenUnlearning](https://github.com/locuslab/open-unlearning), an easy-to-use framework for experimentation supporting multiple unlearning benchmarks (including TOFU and MUSE). Check out the latest documentation for usage details and the new target base models and retain models on [HuggingFace](https://huggingface.co/open-unlearning) across 1B to 8B sizes.
We'd love to hear your feedback and welcome any contributions to the repository!
